src/model/selection.js

```diff
@@ -673,54 +683,27 @@ export default class Selection {
 		for ( const range of this.getRanges() ) {
 			const startBlock = getParentBlock( range.start, visited );
 
-			if ( startBlock ) {
+			if ( startBlock && isTopBlockInRange( startBlock, range ) ) {
 				yield startBlock;
 			}
 
 			for ( const value of range.getWalker() ) {
```

Comment:
As I understand the code below, what happens is that you, at some point, get a block element. But then, the tree walker steps into it and you traverse through more block elements and they might not be the topmost blocks. But you already visited that one, so everything inside can be skipped. This part of code can probably be optimized but the gain wouldn't be big so I won't be pushing on this.

Reply:
Probably the biggest efficiency loss will happen when table is selected. There are a lot of elements in tables and you will have to go through all of them, for nothing.

If you'd like to optimize this, you'd need to change `getWalker()` and use actual `TreeWalker` instance and then use `.skip()` method on it (which is a little difficult to use, so check existing code for examples).

As I said, I won't force it, but it would be nice if the algorithm was a little smarter :P

Reply:
I have the same thought as you (if I read this correctly) that for a table structure the `getSelectedBlocks()` should return: `table, p, p...` but in fact it returns `p, p, table` (children first). So when traversing the tree with a tree walker I don't have that information yet and I must extract it from somewhere).

Previously the `getTopMostBlocks()` would just get all the blocks and then yield only that that aren't inside other blocks which wasn't super great.

Reply:
So no, I havn't visited `table` while visiting paragraph inside table cell.

Reply:
Oh, because you check `elementEnd`, not `elementStart`.

Reply:
Still, the same applies. When you have `<p>Fo[o</p><table>...</table><p>B]ar</p>` selected blocks are `p`, `table`, `p`. So there is no sense to traverse things in the table. But as I said this probably is a minor thing. If you don't want to tinker with it, we can fix it if it proves to be a problem for big tables.

Reply:
OK - we listen there for `elementEnd` only. Maybe because of something of that linked bug in endElement. So we might listen to `elementStart` there and the order will be as expected?

Right now it is `p, p in table, p in table, table, p` because of examing this on `elementEnd` not the `elementStart`. I'll check it briefly.

Reply:
The checks are done on `elementEnd` AFAICS to get only fully selected blocks and to properly check the case of #984 ticket.